rs/bitcoin/kyt/src/state.rs

```diff
@@ -102,27 +95,40 @@ impl<T> FetchTxCache<T> {
     // already have a status, a new status entry is created and associated
     // with the given the timestamp.
     //
-    // When called with non-decreasing timestamps, this function will remove
-    // at most one past (i.e. the oldest) entry based on the `expiry` and
-    // `max_entries` settings, when either condition satisfies.
-    fn set_status_with(&mut self, txid: Txid, status: T, now: Timestamp) {
+    // This function may also remove and return the oldest entry if we
+    // exceeds the `max_entries` settings.
+    fn set_status_with(
+        &mut self,
+        txid: Txid,
+        status: T,
+        now: Timestamp,
+    ) -> Option<(Txid, Timestamp, T)> {
         if self.status.insert(txid, status).is_none() {
             // This is a new entry, record its created time.
             self.created.push_back((txid, now));
-            // Expire the oldest entry if it has already expired, or when we exceed max_entries.
-            if let Some((txid, created_time)) = self.created.front().cloned() {
-                if created_time + self.expiry < now || self.created.len() > self.max_entries {
-                    self.created.pop_front();
-                    self.status.remove(&txid);
-                }
+            // Purge the oldest entry when we exceed max_entries.
+            if self.created.len() > self.max_entries {
+                return self.created.pop_front().and_then(|(txid, timestamp)| {
```

Comment:
nit: before returning maybe add an assert that

```rust
self.created.len() <= self.max_entries
```

Reply:
Good idea! Added the assertion